### DISCUSSION TITLE
Apply explicit type for additionalProperties

### DISCUSSION THREAD

**Discussion by aalmiray** [victools/jsonschema-generator#333] [Q&A] [answered]
Hi, I'm looking for an option that would let me add an explicit `"type": "object"` to the "additionalProperties" element. Reason being that publication of JSON schemas to schemastore requires schemas to be compliant with strict mode https://github.com/SchemaStore/schemastore/issues/2793#issuecomment-1483736023

Basically turn

```json
{
  "additionalProperties": {
    "type": "string"
  }
}
```

into

```json
{
  "type": "object",
  "additionalProperties": {
    "type": "string"
  }
}
```

**Comment by CarstenWickner** [comment]
Hi @aalmiray,

What Schema Draft version are you using?
The referenced example is this:
```json
        "headers" : {
          "allOf" : [ {
            "$ref" : "#/definitions/StringProperties"
          }, {
            "additionalProperties" : {
              "type" : "string"
            }
          } ]
        }
```

However, when you use a newer Schema Draft version (e.g., Draft 2019-09), that supports `$ref` and other keywords side-by-side, the generator should actually clean this up at the end of the process (unless you disabled the `Option.ALLOF_CLEANUP_AT_THE_END`) to this:
```json
        "headers" : {
          "$ref" : "#/definitions/StringProperties"
          "additionalProperties" : {
            "type" : "string"
          }
        }
```

**Reply by aalmiray** [reply]
I'm using DRAFT_7 as I believe schemastore prefers older versions for compatibility reasons https://github.com/jreleaser/jreleaser/blob/10742852a58603b852c6debafb9d50034d956b58/core/jreleaser-engine/src/main/java/org/jreleaser/engine/schema/JsonSchemaGenerator.java#L104

**Reply by aalmiray** [reply]
I've tried using Draft 2019-09 and `Option.ALLOF_CLEANUP_AT_THE_END` and I can see big changes in the generated schema. Sadly schemastore doesn't like this draft at all. 

```
################ Error message
>> Schema version is too high => "2019-09" in file jreleaser-1.5.0.json
>> Schema version 2019-09,2020-12 is not supported by many editors and IDEs
>> jreleaser-1.5.0.json must use a lower schema version.
##############################
```

`DRAFT_7` and `Option.ALLOF_CLEANUP_AT_THE_END` has no effect.

**Reply by CarstenWickner** [reply]
The `allOf` clean-up is enabled by default, that's why you don't see a difference under Draft 7.

Up to Draft 7, the `$ref` keyword caused any other keyword in the same subschema to be ignored. That's the reason for `allOf` being used there.
In newer drafts, that limitation was removed.

I'm currently working on a new `Option`, that should allow you to satisfy this "strict" type indication.
Let's see whether I'm successful. 😅

**Reply by CarstenWickner** [reply]
I've got a draft now: https://github.com/victools/jsonschema-generator/pull/334
That produces this then (since it'd be more complicated to find out whether another "allOf" part allows `null` values – especially if a potentially external `$ref` is used):
```json
{
  "type": ["object", "null"],
  "additionalProperties": {
    "type": "string"
  }
}
```

What do you reckon?

**Reply by aalmiray** [reply]
Nice! I'll try manually adding that snippet and run schemastore's checker again.

**Reply by aalmiray** [reply]
I can confirm that adding the type annotation to _every_ single place where `additionalProperties` has a struct value makes schemastore happy, that is

```json
      "type": ["object", "null"],
      "additionalProperties" : {
        "type" : "string"
      }
```

**Reply by CarstenWickner** [reply]
I've just pushed a `4.30.0-SNAPSHOT` version to Maven (including the new `Option`), in case you want to try that one out as well.
Not sure when I will get to the release.

Do you have a particular urgency/time pressure here, or is it acceptable not to release this yet?

**Reply by aalmiray** [reply]
Thanks! It's not urgent at the moment as JReleaser's schema is independently hosted somewhere else, also available on demand via its CLI. Having it at schemastore is the cherry in top.

**Reply by CarstenWickner** [reply, marked as answer]
Release [v4.30.0](https://github.com/victools/jsonschema-generator/releases/tag/v4.30.0) has been published. 🚀

The new `Option.STRICT_TYPE_INFO` can be used to achieve the desired output here.
